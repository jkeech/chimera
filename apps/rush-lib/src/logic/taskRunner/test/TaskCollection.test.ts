// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TaskCollection } from '../TaskCollection';
import { Task } from '../Task';
import { StringBufferTerminalProvider } from '@rushstack/node-core-library';
import { MockBuilder } from './MockBuilder';
import { TaskStatus } from '../TaskStatus';

function checkConsoleOutput(terminalProvider: StringBufferTerminalProvider): void {
  expect(terminalProvider.getOutput()).toMatchSnapshot();
  expect(terminalProvider.getVerbose()).toMatchSnapshot();
  expect(terminalProvider.getWarningOutput()).toMatchSnapshot();
  expect(terminalProvider.getErrorOutput()).toMatchSnapshot();
}

describe('TaskCollection', () => {
  let terminalProvider: StringBufferTerminalProvider;
  let taskCollection: TaskCollection;

  beforeEach(() => {
    terminalProvider = new StringBufferTerminalProvider(true);
  });

  describe('Dependencies', () => {
    beforeEach(() => {
      taskCollection = new TaskCollection();
    });

    it('throwsErrorOnNonExistentTask', () => {
      expect(() => taskCollection.addDependencies('foo', [])).toThrowErrorMatchingSnapshot();
    });

    it('throwsErrorOnNonExistentDependency', () => {
      taskCollection.addTask(new MockBuilder('foo'));
      expect(() => taskCollection.addDependencies('foo', ['bar'])).toThrowErrorMatchingSnapshot();
    });

    it('detectsDependencyCycle', () => {
      taskCollection.addTask(new MockBuilder('foo'));
      taskCollection.addTask(new MockBuilder('bar'));
      taskCollection.addDependencies('foo', ['bar']);
      taskCollection.addDependencies('bar', ['foo']);
      expect(() => taskCollection.getOrderedTasks()).toThrowErrorMatchingSnapshot();
    });

    it('respectsDependencyOrder', () => {
      const result: string[] = [];
      taskCollection.addTask(
        new MockBuilder('two', async () => {
          result.push('2');
          return TaskStatus.Success;
        })
      );
      taskCollection.addTask(
        new MockBuilder('one', async () => {
          result.push('1');
          return TaskStatus.Success;
        })
      );
      taskCollection.addDependencies('two', ['one']);

      const tasks: Task[] = taskCollection.getOrderedTasks();
      expect(tasks.map((t) => t.name).join(',')).toEqual('one,two');
      checkConsoleOutput(terminalProvider);
    });
  });

  describe('Error logging', () => {
    beforeEach(() => {
      taskCollection = new TaskCollection();
    });
  });
});
