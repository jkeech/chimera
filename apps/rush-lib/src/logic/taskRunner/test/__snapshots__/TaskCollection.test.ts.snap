// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCollection Dependencies detectsDependencyCycle 1`] = `
"A cyclic dependency was encountered:
  foo
  -> bar
  -> foo
Consider using the cyclicDependencyProjects option for rush.json."
`;

exports[`TaskCollection Dependencies respectsDependencyOrder 1`] = `""`;

exports[`TaskCollection Dependencies respectsDependencyOrder 2`] = `""`;

exports[`TaskCollection Dependencies respectsDependencyOrder 3`] = `""`;

exports[`TaskCollection Dependencies respectsDependencyOrder 4`] = `""`;

exports[`TaskCollection Dependencies throwsErrorOnNonExistentDependency 1`] = `"The project 'bar' has not been registered."`;

exports[`TaskCollection Dependencies throwsErrorOnNonExistentTask 1`] = `"The task 'foo' has not been registered"`;
